staking/parse.go

```diff
@@ -3,13 +3,161 @@
 
 package staking
 
-import "crypto/x509"
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"golang.org/x/crypto/cryptobyte"
+
+	cryptobyte_asn1 "golang.org/x/crypto/cryptobyte/asn1"
+)
+
+var (
+	ErrMalformedCertificate                  = errors.New("staking: malformed certificate")
+	ErrMalformedTBSCertificate               = errors.New("staking: malformed tbs certificate")
+	ErrMalformedVersion                      = errors.New("staking: malformed version")
+	ErrMalformedSerialNumber                 = errors.New("staking: malformed serial number")
+	ErrMalformedSignatureAlgorithmIdentifier = errors.New("staking: malformed signature algorithm identifier")
+	ErrMalformedIssuer                       = errors.New("staking: malformed issuer")
+	ErrMalformedValidity                     = errors.New("staking: malformed validity")
+	ErrMalformedSPKI                         = errors.New("staking: malformed spki")
+	ErrMalformedPublicKeyAlgorithmIdentifier = errors.New("staking: malformed public key algorithm identifier")
+	ErrMalformedSubjectPublicKey             = errors.New("staking: malformed subject public key")
+	ErrMalformedOID                          = errors.New("staking: malformed oid")
+	ErrInvalidRSAModulus                     = errors.New("staking: invalid RSA modulus")
+	ErrInvalidRSAPublicExponent              = errors.New("staking: invalid RSA public exponent")
+	ErrRSAModulusNotPositive                 = errors.New("staking: RSA modulus is not a positive number")
+	ErrRSAPublicExponentNotPositive          = errors.New("staking: RSA public exponent is not a positive number")
+	ErrFailedUnmarshallingEllipticCurvePoint = errors.New("staking: failed to unmarshal elliptic curve point")
+	ErrUnknownPublicKeyAlgorithm             = errors.New("staking: unknown public key algorithm")
+)
 
 // ParseCertificate parses a single certificate from the given ASN.1 DER data.
+//
+// TODO: Remove after v1.11.x activates.
 func ParseCertificate(der []byte) (*Certificate, error) {
-	cert, err := x509.ParseCertificate(der)
+	x509Cert, err := x509.ParseCertificate(der)
 	if err != nil {
 		return nil, err
 	}
-	return CertificateFromX509(cert), nil
+	stakingCert := CertificateFromX509(x509Cert)
+	return stakingCert, ValidateCertificate(stakingCert)
+}
+
+// ParseCertificatePermissive parses a single certificate from the given ASN.1.
+//
+// This function does not validate that the certificate is validate to be used
+// against normal TLS implementations.
+//
+// Ref: https://github.com/golang/go/blob/go1.19.12/src/crypto/x509/parser.go#L789-L968
+func ParseCertificatePermissive(bytes []byte) (*Certificate, error) {
+	if len(bytes) > MaxCertificateLen {
+		return nil, ErrCertificateTooLarge
+	}
+
+	input := cryptobyte.String(bytes)
+	// Consume the length and tag bytes.
+	if !input.ReadASN1(&input, cryptobyte_asn1.SEQUENCE) {
```

Comment:
Technically - there is no reason to enforce the `tag`s here. Should we switch to using `ReadAnyASN1`? It might be nice to keep the tags just to make it more clear what we are reading/skipping. Although it would be faster to just ignore the tag.

Reply:
Behind the scenes, this just calls `ReadAnyASN1` and compares the tag afterwards. If speed is a factor in the decision, I don't think that switch would actually make much of an impact: https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.16.0:cryptobyte/asn1.go;l=594-605